dex/src/main/java/io/crate/data/FilteringBatchIterator.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import java.util.function.Predicate;
+
+public class FilteringBatchIterator extends ForwardingBatchIterator {
+
+    private final Predicate<Row> filter;
+
+    public FilteringBatchIterator(BatchIterator delegate, Predicate<Row> filter) {
+        super(delegate);
+        this.filter = filter;
+    }
+
+    @Override
+    public boolean moveNext() {
+        while (super.moveNext()) {
+            if (filter.test(super.currentRow())) {
```

Comment:
it was required, otherwise it would have been a recursion. I changed it to make the delegate call more explicit.